core/src/main/java/org/apache/calcite/rel/core/Project.java

```diff
@@ -89,25 +94,33 @@ protected Project(
       List<RelHint> hints,
       RelNode input,
       List<? extends RexNode> projects,
-      RelDataType rowType) {
+      RelDataType rowType,
+      Set<CorrelationId> variablesSet) {
     super(cluster, traits, input);
     assert rowType != null;
     this.exps = ImmutableList.copyOf(projects);
     this.hints = ImmutableList.copyOf(hints);
     this.rowType = rowType;
+    this.variablesSet = Objects.requireNonNull(
+        ImmutableSet.copyOf(variablesSet), "variablesSet");
```

Comment:
Can `copyOf` ever return `null`?

Reply:
No, it's an artifact of moving this check to parent class :(. `ImmutableSet.copyOf` will throw an NPE if someone will pass a null parameter. So we need to wrap an argument `variablesSet` with `requiredNonNull`, and pass the result to `copyOf`.

Thanks for noticing! Fixed.